File: modules/config/crypto.go
package config

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"
	"io"
)

const nonceSize = 12

// Encrypt encrypts plaintext using AES-256-GCM with the given 32-byte key.
// The 12-byte nonce is prepended to the ciphertext.
func Encrypt(key, plaintext []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, fmt.Errorf("key must be 32 bytes, got %d", len(key))
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("create AES cipher: %w", err)
	}

	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("create GCM: %w", err)
	}

	nonce := make([]byte, nonceSize)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, fmt.Errorf("generate nonce: %w", err)
	}

	// Seal appends ciphertext+tag to nonce
	ciphertext := aead.Seal(nonce, nonce, plaintext, nil)
	return ciphertext, nil
}

// Decrypt decrypts ciphertext produced by Encrypt using the given 32-byte key.
// It reads the nonce from the first 12 bytes of ciphertext.
func Decrypt(key, ciphertext []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, fmt.Errorf("key must be 32 bytes, got %d", len(key))
	}

	if len(ciphertext) < nonceSize+1 {
		return nil, fmt.Errorf("ciphertext too short")
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("create AES cipher: %w", err)
	}

	aead, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("create GCM: %w", err)
	}

	nonce := ciphertext[:nonceSize]
	data := ciphertext[nonceSize:]

	plaintext, err := aead.Open(nil, nonce, data, nil)
	if err != nil {
		return nil, fmt.Errorf("decrypt: %w", err)
	}

	return plaintext, nil
}

File: packages/config/src/crypto.ts
import { newConfigError } from "@orangecoding/core";

const NONCE_SIZE = 12;

/**
 * Convert a Uint8Array to a BufferSource that is compatible with the Web
 * Crypto API's strict type requirements for ArrayBuffer (not SharedArrayBuffer).
 */
function toBufferSource(data: Uint8Array): ArrayBuffer {
  // Create a plain ArrayBuffer copy to satisfy the Web Crypto API's
  // strict type requirement (ArrayBuffer, not SharedArrayBuffer).
  const buf = new ArrayBuffer(data.byteLength);
  new Uint8Array(buf).set(data);
  return buf;
}

/**
 * Encrypt encrypts plaintext using AES-256-GCM with the given 32-byte key.
 * The 12-byte nonce is prepended to the ciphertext.
 */
export async function encrypt(
  key: Uint8Array,
  plaintext: Uint8Array,
): Promise<Uint8Array> {
  if (key.length !== 32) {
    throw newConfigError(`key must be 32 bytes, got ${key.length}`);
  }

  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_SIZE));
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    toBufferSource(key),
    { name: "AES-GCM" },
    false,
    ["encrypt"],
  );

  const encrypted = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonce, tagLength: 128 },
    cryptoKey,
    toBufferSource(plaintext),
  );

  // Prepend nonce to ciphertext (nonce | ciphertext+tag)
  const result = new Uint8Array(NONCE_SIZE + encrypted.byteLength);
  result.set(nonce, 0);
  result.set(new Uint8Array(encrypted), NONCE_SIZE);

  return result;
}

/**
 * Decrypt decrypts ciphertext produced by encrypt using the given 32-byte key.
 * It reads the nonce from the first 12 bytes of ciphertext.
 */
export async function decrypt(
  key: Uint8Array,
  ciphertext: Uint8Array,
): Promise<Uint8Array> {
  if (key.length !== 32) {
    throw newConfigError(`key must be 32 bytes, got ${key.length}`);
  }

  if (ciphertext.length < NONCE_SIZE + 1) {
    throw newConfigError("ciphertext too short");
  }

  const nonce = ciphertext.slice(0, NONCE_SIZE);
  const data = ciphertext.slice(NONCE_SIZE);

  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    toBufferSource(key),
    { name: "AES-GCM" },
    false,
    ["decrypt"],
  );

  try {
    const plaintext = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: nonce, tagLength: 128 },
      cryptoKey,
      toBufferSource(data),
    );
    return new Uint8Array(plaintext);
  } catch (err) {
    throw newConfigError(
      `decrypt: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}
